hack/lib/util/dind.sh

```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
+#
+# This library holds utility functions used by dind deployment and images.
+
+# os::util::wait-for-condition blocks until the provided condition becomes true
+#
+# Globals:
+#  None
+# Arguments:
+#  - 1: message indicating what conditions is being waited for (e.g. 'config to be written')
+#  - 2: a string representing an eval'able condition.  When eval'd it should not output
+#       anything to stdout or stderr.
+#  - 3: optional timeout in seconds.  If not provided, defaults to 60s.  If OS_WAIT_FOREVER
+#       is provided, wait forever.
+# Returns:
+#  1 if the condition is not met before the timeout
+readonly OS_WAIT_FOREVER=-1
+function os::util::wait-for-condition() {
+  local msg=$1
+  # condition should be a string that can be eval'd.  When eval'd, it
+  # should not output anything to stderr or stdout.
+  local condition=$2
+  local timeout=${3:-60}
+
+  local start_msg="Waiting for ${msg}"
+  local error_msg="[ERROR] Timeout waiting for ${msg}"
+
+  local counter=0
+  while ! ${condition}; do
+    if [[ "${counter}" = "0" ]]; then
+      echo "${start_msg}"
+    fi
+
+    if [[ "${counter}" -lt "${timeout}" ||
+            "${timeout}" = "${OS_WAIT_FOREVER}" ]]; then
+      counter=$((counter + 1))
+      if [[ "${timeout}" != "${OS_WAIT_FOREVER}" ]]; then
+        echo -n '.'
+      fi
+      sleep 1
+    else
+      echo -e "\n${error_msg}"
+      return 1
+    fi
+  done
+
+  if [[ "${counter}" != "0" && "${timeout}" != "${OS_WAIT_FOREVER}" ]]; then
+    echo -e '\nDone'
+  fi
+}
```

Comment:
to match our `import`s in `hack/lib/init.sh` we always declare our functions `readonly` -- otherwise there is a chance that a nested script with nested callouts to `hack/lib/init.sh` will have a different version of a function, or an alias for it or other such nonsense, which we don't really want to deal with